collect_app/src/test/java/org/odk/collect/android/preferences/AdminPasswordDialogFragmentTest.java

```diff
@@ -0,0 +1,108 @@
+import org.odk.collect.android.support.TestActivityScenario;
+import org.odk.collect.android.utilities.AdminPasswordProvider;
+import org.robolectric.annotation.LooperMode;
+
+import static android.os.Looper.getMainLooper;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+import static org.odk.collect.android.preferences.AdminPasswordDialogFragment.Action.STORAGE_MIGRATION;
+import static org.robolectric.Shadows.shadowOf;
+import static org.robolectric.annotation.LooperMode.Mode.PAUSED;
+
+@RunWith(AndroidJUnit4.class)
+@LooperMode(PAUSED)
+public class AdminPasswordDialogFragmentTest {
+
+    @Before
+    public void setup() {
+        RobolectricHelpers.overrideAppDependencyModule(new AppDependencyModule() {
+            @Override
+            public AdminPasswordProvider providesAdminPasswordProvider() {
+                return new StubAdminPasswordProvider();
+            }
+        });
+    }
+
+    @Test
+    public void enteringPassword_andClickingOK_callsOnCorrectAdminPasswordWithAction() {
+        TestActivityScenario<MockAdminPasswordDialogCallback> activityScenario = TestActivityScenario.launch(MockAdminPasswordDialogCallback.class);
+        activityScenario.onActivity(activity -> {
+            AdminPasswordDialogFragment fragment = new AdminPasswordDialogFragment();
+            fragment.setAction(STORAGE_MIGRATION);
+            fragment.show(activity.getSupportFragmentManager(), "tag");
+            shadowOf(getMainLooper()).idle();
+
+            fragment.getInput().setText("password");
+            ((AlertDialog) fragment.getDialog()).getButton(AlertDialog.BUTTON_POSITIVE).performClick();
+            shadowOf(getMainLooper()).idle();
+
+            assertThat(activity.onCorrectAdminPasswordCalledWith, equalTo(STORAGE_MIGRATION));
+        });
+    }
+
+    @Test
+    public void afterRecreating_enteringPassword_andClickingOK_callsOnCorrectAdminPasswordWithAction() {
+        TestActivityScenario<MockAdminPasswordDialogCallback> activityScenario = TestActivityScenario.launch(MockAdminPasswordDialogCallback.class);
+        activityScenario.onActivity(activity -> {
+            AdminPasswordDialogFragment fragment = new AdminPasswordDialogFragment();
+            fragment.setAction(STORAGE_MIGRATION);
+            fragment.show(activity.getSupportFragmentManager(), "tag");
+            shadowOf(getMainLooper()).idle();
+        });
+
+        activityScenario.recreate();
+        activityScenario.onActivity(activity -> {
+            AdminPasswordDialogFragment fragment = (AdminPasswordDialogFragment) activity.getSupportFragmentManager().findFragmentByTag("tag");
+
+            fragment.getInput().setText("password");
+            ((AlertDialog) fragment.getDialog()).getButton(AlertDialog.BUTTON_POSITIVE).performClick();
+            shadowOf(getMainLooper()).idle();
+
+            assertThat(activity.onCorrectAdminPasswordCalledWith, equalTo(STORAGE_MIGRATION));
```

Comment:
I can't wrap my brain around why you can call a private `MockAdminPasswordDialogCallback` field here?

Reply:
Because in `onAttach` we have:

```java
if (context instanceof AdminPasswordDialogCallback) {
    callback = (AdminPasswordDialogCallback) context;
}
```

The dialog assumes expects its host activity to implement the listener. We could decouple this but I didn't want to get into that. I'm not a fan of the pattern as it's awkward for testing (which is always a sign that the design is a little awkward) but it is actually how the Android [docs suggest this should be done](https://developer.android.com/guide/components/fragments#EventCallbacks). Using an actual listener object would be a problem as it could cause weirdness with Fragment recreation.